repo-service/src/clojure/org/cyverse/dataone/DataOneRepoService.clj

```diff
@@ -84,17 +92,122 @@
+        (when (pos? limit)
+          (try
+            (.getFirstResult rs)
+            (catch DataNotFoundException _ nil))))
+
+      ;; It's sufficient to return the number of columns from the original result set.
+      (getNumberOfResultColumns [_]
+        (.getNumberOfResultColumns rs))
+
+      ;; The set of results should be empty if the limit is zero.
+      (getResults [_]
+        (if (pos? limit)
+          (vec (.getResults rs))
+          []))
+
+      ;; Generic queries don't return the total number of matching records if the offset is greater than the total
+      ;; number of matching records. The workaround that we have for this is to perform a second query with an offset
+      ;; of zero any time the offset is positive and the total record count isn't.
+      (getTotalRecords [_]
+        (.getTotalRecords count-rs))
+
+      ;; If the limit is zero then the original result set's isHasMoreRecords method will still return a false
+      ;; value if there was exactly one more matching record.
+      (isHasMoreRecords [_]
+        (if (pos? limit)
+          (.isHasMoreRecords rs)
+          (pos? (count (.getResults rs))))))))
+
+(defn- run-queries
+  "Runs multiple generic queries and returns a sequence of result sets. This is a workaround for iRODS generic
+   queries not supporting `or` conditions. The columns returned by all queries passed to this function must be
+   identical."
+  [this query-builders offset limit]
+  (loop [[builder & builders] query-builders
+         offset               (max offset 0)
+         limit                (max limit 0)
+         acc                  []]
+    (if builder
+      (let [rs    (run-query this builder offset limit)
+            size  (count (.getResults rs))
+            total (.getTotalRecords rs)]
+        (if (= size 0)
+          (recur builders (max (- offset total) 0) limit (conj acc rs))
+          (recur builders 0 (max (- limit size) 0) (conj acc rs))))
+      acc)))
+
+(defn- gen-queries
+  "Performs multiple generic queries and returns a combined result set. The result set will be closed automatically.
+   The columns returned by all queries passed to this function must be identical."
+  [this query-builders offset limit]
+  (let [rss (run-queries this query-builders offset limit)]
+    (reify
+      IRODSQueryResultSetInterface
+
+      ;; Since we're assuming that all queries return the same set of columns, this method only returns the column
+      ;; names of the first result set.
+      (getColumnNames [_]
+        (.getColumnNames (first rss)))
+
+      ;; The first result set may not contain any records, so it's necessary to check every result set.
+      (getFirstResult [this]
+        (first (.getResults this)))
+
+      ;; Since we're assuming that all queries return the same set of columns, this method only returns the number of
+      ;; columns in the first result set.
+      (getNumberOfResultColumns [_]
+        (.getNumberOfResultColumns (first rss)))
```

Comment:
Possibly a dumb question, but if the first result set has no records, will this method (and I guess `getColumnNames` as well) still return something?

Reply:
Awesome, thanks for checking!  LGTM